x/stakeibc/keeper/unbonding.go

```diff
@@ -400,57 +455,28 @@ func (k Keeper) UnbondFromHostZone(ctx sdk.Context, hostZone types.HostZone) (er
 	}
 
 	// Send the messages in batches so the gas limit isn't exceedeed
-	for start := 0; start < len(msgs); start += undelegateBatchSize {
-		end := start + undelegateBatchSize
-		if end > len(msgs) {
-			end = len(msgs)
-		}
-
-		msgsBatch := msgs[start:end]
-		unbondingsBatch := unbondings[start:end]
-
-		// Store the callback data
-		undelegateCallback := types.UndelegateCallback{
-			HostZoneId:              hostZone.ChainId,
-			SplitUndelegations:      unbondingsBatch,
-			EpochUnbondingRecordIds: epochUnbondingRecordIds,
-		}
-		callbackArgsBz, err := proto.Marshal(&undelegateCallback)
-		if err != nil {
-			return errorsmod.Wrap(err, "unable to marshal undelegate callback args")
-		}
-
-		// Submit the undelegation ICA
-		if _, err := k.SubmitTxsDayEpoch(
-			ctx,
-			hostZone.ConnectionId,
-			msgsBatch,
-			types.ICAAccountType_DELEGATION,
-			ICACallbackID_Undelegate,
-			callbackArgsBz,
-		); err != nil {
-			return errorsmod.Wrapf(err, "unable to submit unbonding ICA for %s", hostZone.ChainId)
-		}
-
-		// flag the delegation change in progress on each validator
-		for _, unbonding := range unbondingsBatch {
-			if err := k.IncrementValidatorDelegationChangesInProgress(&hostZone, unbonding.Validator); err != nil {
-				return err
-			}
-		}
-		k.SetHostZone(ctx, hostZone)
-	}
-
-	// Update the epoch unbonding record status
-	if err := k.RecordsKeeper.SetHostZoneUnbondingStatus(
+	numTxsSubmitted, err := k.BatchSubmitUndelegateICAMessages(
 		ctx,
-		hostZone.ChainId,
+		hostZone,
 		epochUnbondingRecordIds,
-		recordstypes.HostZoneUnbonding_UNBONDING_IN_PROGRESS,
-	); err != nil {
+		msgs,
+		unbondings,
+		undelegateBatchSize,
+	)
+	if err != nil {
 		return err
 	}
 
+	// Update the epoch unbonding record status and number of undelegation ICAs
+	for epochNumber, hostZoneUnbonding := range epochNumberToHostZoneUnbonding {
+		hostZoneUnbonding.Status = recordstypes.HostZoneUnbonding_UNBONDING_IN_PROGRESS
+		hostZoneUnbonding.UndelegationTxsInProgress += numTxsSubmitted
```

Comment:
Now that things are batched, this numTxsSubmitted isn't really specific to the this particular hostZoneUnbonding record but I see how it will function largely the same, I assume in the callback the plan is to decrement this number to ensure that all relevant Txs have finished and since all will succeed or fail together, I guess this is still reasonable to apply it here on each hostZoneUnbonding record but it did cause me to pause on reading it the first time.

Reply:
yeah, this is where it's a bit messy to have messages overlap records. But the main reason for this is so that we know to only initiate a retry when there's no more outstanding messages

Otherwise you could have a scenario where the last batch lands on the host zone in time (so there's no timeout) but the ack doesn't come back until after the next epoch. In this case we would accidentally re-submit that failed batch

Reply:
Good looking out for that edge case.  In general I think these semaphores are the more robust way to handle the inherent race conditions without having to think through every case -- definitely a good pattern I think we had done during LSM the first time that I noticed it in the codebase.